### PR TITLE
Implement PerformanceCalculator for osu! to calculate PP

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/1516

Similar implementation to DifficultyCalculator. Uses SR generated by a DifficultyCalculator, and then applies some scaling factors to determine the final value.

As far as I can see (multiple passes) this is equivalent to what is inside osu-stable's calculator, but I'm not getting the same pp values, so I'm blocking this PR for now because of that.

There is a TestCase which displays SR + online PP + local PP which demonstrates this discrepancy. Also, it seems like DT does not really affect SR much (or, even negatively affects SR), @Tom94 any ideas?

### CODE REVIEW

**Comment by Tom94** [comment]
I recommend you check the existing implementation of mods in osu-stable to ensure they are treated the exact same way. Your comment about SR not changing with DT suggests, that the map is not being correctly transformed, which may as well be true for HR. There are some subtleties in how mods need to be masked, too.

**Comment by smoogipoo** [comment]
At this point I believe the SR calculations are close enough - within 1d.p. at least, which is probably attributed to various cumulative inaccuracies vs stable...

**Comment by smoogipoo** [comment]
PP calculation is also as close as it's going to get right now - bar inaccuracies resulting from DT not modifying AR and the calculator utilising values rounded to 2d.p. from DifficultyCalculators.

**Comment by Tom94** [comment]
1d.p. is pretty large for SR, though. Does it match up better than 1d.p. for nomod?

**Comment by Tom94** [comment]
DT not modifying AR and using 2d.p. rounded values from `DifficultyCalculator` sound easily fixable. Why not address them within this PR?

**Comment by smoogipoo** [comment]
@Tom94 This is as close as I can get it for now. It's still off, but I have not had much luck as to why that is the case. If you have the time, please look through it and sanity check it.

**Comment by peppy** [comment]
Can we remove the blocked label from this one? also is pending on some CI fixes @smoogipoo.

**Comment by peppy** [comment]
Can we disable the failing test to make forward progress on this? I've already gone through the code and from a stylistic perspective I'm happy with it.

File: osu.Game.Rulesets.Osu/Scoring/OsuPerformanceCalculator.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            mods = Score.Mods;
+            accuracy = Score.Accuracy;
+            scoreMaxCombo = Score.MaxCombo;
+            count300 = Convert.ToInt32(Score.Statistics["300"]);
+            count100 = Convert.ToInt32(Score.Statistics["100"]);
+            count50 = Convert.ToInt32(Score.Statistics["50"]);
+            countMiss = Convert.ToInt32(Score.Statistics["x"]);
+
+            // Don't count scores made with supposedly unranked mods
+            if (mods.Any(m => m is OsuModRelax || m is OsuModAutopilot || m is OsuModAutoplay))
```
Comment (Aergwyn) [inline osu.Game.Rulesets.Osu/Scoring/OsuPerformanceCalculator.cs]: I have a question out of curiousity: Why not `mods.Any(m => !m.Ranked)`?

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Scoring/OsuPerformanceCalculator.cs]: Good point.

**Review by peppy — APPROVED** [review]
calculations don't match yet, but let's get this in as it's better than what we had (and provides a solid base for further testing).

**Pull request merged**